libraries/lib-ipc/internal/BufferedIPCChannel.cpp

```diff
@@ -0,0 +1,158 @@
+#include "BufferedIPCChannel.h"
+
+#include <cassert>
+#include <cstring>
+#include <mutex>
+#include "MemoryX.h"
+#include "socket_guard.h"
+
+BufferedIPCChannel::BufferedIPCChannel()
+{
+   mOutputBuffer.reserve(DefaultOutputBufferCapacity);
+}
+
+BufferedIPCChannel::~BufferedIPCChannel()
+{
+   if(mSocket != INVALID_SOCKET)
+      //interrupts any blocking calls
+      CLOSE_SOCKET(mSocket);
+
+   if(mSendRoutine)
+      mSendRoutine->join();
+   if(mRecvRoutine)
+      mRecvRoutine->join();
+}
+
+void BufferedIPCChannel::Send(const void* bytes, size_t length)
+{
+   std::lock_guard lck(mSocketSync);
+
+   auto offset = mOutputBuffer.size();
+   mOutputBuffer.resize(offset + length);
+   memcpy(mOutputBuffer.data() + offset, bytes, length);
+
+   mSendCondition.notify_one();
```

Comment:
There are two uses of `notify_one`.  It is explained here that the notifier does not need to lock the mutex and it might be slightly bad for performance if it does so. https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one

Reply:
In this case mutual exclusion is needed for other reasons of course.  Still the appending to the buffer must be excluded from the consuming of it.